pkg/internalregistry/controllers/image_pull_secret_controller.go

```diff
@@ -329,29 +330,35 @@ CheckUrl:
 		}
 		if !validKeyID {
 			klog.V(2).InfoS("Internal registry pull secret needs to be refreshed", "reason", "auth token was not signed by a current signer", "ns", imagePullSecret.Namespace, "name", imagePullSecret.Name, "url", url, "error", err)
-			return false, time.Now()
+			return false, nil
 		}
 
 		var claims jwt.Claims
 		// "unsafe" in the following API just means we are not validating the signature
 		err = token.UnsafeClaimsWithoutVerification(&claims)
 		if err != nil {
 			klog.V(2).InfoS("Internal registry pull secret needs to be refreshed", "reason", "auth token claim cannot be parsed", "ns", imagePullSecret.Namespace, "name", imagePullSecret.Name, "url", url, "error", err)
-			return false, time.Now()
+			return false, nil
 		}
-		// if token is expired or will only be valid less than 40% of its remaining duration we want to trigger a new token request
-		refreshTime := claims.Expiry.Time().Add(time.Duration(-int64(float64(time.Now().Sub(claims.Expiry.Time())) * 0.4)))
-		klog.V(4).InfoS("Token expiration check.", "ns", imagePullSecret.Namespace, "name", imagePullSecret.Name, "url", url, "expirtyTime", claims.Expiry.Time(), "refreshTime", refreshTime)
-		if time.Now().After(refreshTime) {
-			klog.V(2).InfoS("Internal registry pull secret needs to be refreshed", "reason", "auth token needs to be refreshed", "ns", imagePullSecret.Namespace, "name", imagePullSecret.Name, "url", url, "expirtyTime", claims.Expiry.Time(), "refreshTime", refreshTime)
-			return false, time.Now()
+		// if token is expired or less than 40% of its valid duration is left, we want to trigger a new token request
+		refreshTime := refreshThresholdTime(claims.NotBefore.Time(), claims.Expiry.Time())
+		klog.V(4).InfoS("Token expiration check.", "ns", imagePullSecret.Namespace, "name", imagePullSecret.Name, "url", url, "exp", claims.Expiry.Time(), "refreshTime", refreshTime)
+		if now.After(refreshTime) {
+			klog.V(2).InfoS("Internal registry pull secret needs to be refreshed", "reason", "auth token needs to be refreshed", "ns", imagePullSecret.Namespace, "name", imagePullSecret.Name, "url", url, "exp", claims.Expiry.Time(), "refreshTime", refreshTime)
+			return false, nil
 		}
 		if requeueAt.IsZero() || requeueAt.After(refreshTime) {
 			requeueAt = refreshTime
 		}
 	}
 	klog.V(4).InfoS("Internal registry pull secret does not need to be refreshed.", "ns", imagePullSecret.Namespace, "name", imagePullSecret.Name)
-	return true, requeueAt
+	return true, &requeueAt
+}
+
+func refreshThresholdTime(nbf, exp time.Time) time.Time {
+	// calculate the time at which only 40% of the valid duration would be left
+	validDuration := exp.Sub(nbf)
+	return exp.Add(-time.Duration(int64(float64(validDuration) * 0.4)))
```

Comment:
Are we guarantee to always get `exp` > `nbf` here ?

Reply:
It would be very unlikely, it means k8s would have issued an invalid token. I added a check anyway.